### PR TITLE
winpcap: add winpcap/4.1.3 recipe

### PR DESCRIPTION
**Pull request by madebr**
Specify library name and version:  **winpcap/4.1.3**

- Tested on Linux
- On Windows:
   - Needs the Windows DDK (=Driver Development Kit)
   - Fails to build on my machine because of a missing header inside the DDK (MSVC cannot find `ndis/types.h`). Feedback welcome.
   - Not tested with mingw
- other operating systems: not tested. The recipe raises an error. PRs are welcome to add macos support.

Closes https://github.com/bincrafters/community/pull/1395

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by SpaceIm** [comment]
Did you write all the CMakeLists from scratch or does it come from a fork?

**Comment by madebr** [comment]
I wrote them from scratch.
FindWDK.cmake is the only file I copied from the interwebs. The file head its license at the top.

**Comment by madebr** [comment]
I need to investigate why Windows is not built on cci.

File: recipes/winpcap/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+            self.options.turbocap = True
+            self.options.capture_type = "win32"
+        elif self.settings.os == "Linux":
+            self.options.capture_type = "linux"
+        if self.settings.os != "Linux":
+            del self.options.bluetooth
+            del self.options.usb_sniffing
+            del self.options.usb_sniffing_device
```
Comment (madebr) [inline recipes/winpcap/all/conanfile.py:58]: @jgsogo 
c3i fails to build on MSVC@Windows.
See https://github.com/conan-io/conan-center-index/pull/6348#issuecomment-880414443
c3i says my recipe emits an `ConanInvalidConfiguration` when `os=Windows`.

The options inside `default_options` are indeed invalid on Windows,
but they are converted to valid options in `config_options`.

File: recipes/winpcap/all/cmake/packetntx.cmake [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        PRIVATE
+            source_subfolder/packetNtx/driver/tme.c
+            source_subfolder/packetNtx/driver/count_packets.c
+            source_subfolder/packetNtx/driver/tcp_session.c
+            source_subfolder/packetNtx/driver/functions.c
+            source_subfolder/packetNtx/driver/bucket_lookup.c
+            source_subfolder/packetNtx/driver/normal_lookup.c
+            source_subfolder/packetNtx/driver/win_bpf_filter_init.c
+    )
+    target_compile_definitions(driver PRIVATE HAVE_BUGGY_TME_SUPPORT)
```
Comment (madebr) [inline recipes/winpcap/all/cmake/packetntx.cmake:45]:
```suggestion
    target_compile_definitions(packet PRIVATE HAVE_BUGGY_TME_SUPPORT)
```

File: recipes/winpcap/all/test_package/test_package.c [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+#include "pcap.h"
+
+#include <stdlib.h>
+#include <time.h>
+
+/* prototype of the packet handler */
+void packet_handler(u_char *param, const struct pcap_pkthdr *header, const u_char *pkt_data);
+
+int main()
```
Comment (madebr) [inline recipes/winpcap/all/test_package/test_package.c:11]: This test package is interactive. It should be made non-interactive.

File: recipes/winpcap/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        if self.options.turbocap and self.settings.os != "Windows":
+            # FIXME: missing turbocap recipe
+            raise ConanInvalidConfiguration("missing turbocap cci recipe")
+
+    def build_requirements(self):
+        if tools.os_info.is_windows:
+            self.build_requires("winflexbison/2.5.24")
+        else:
+            self.build_requires("bison/3.7.1")
+            self.build_requires("flex/2.6.4")
```
Comment (SpaceIm) [inline recipes/winpcap/all/conanfile.py:80]: here is the issue with windows.

Comment (madebr) [reply recipes/winpcap/all/conanfile.py:80]: But it's a build requirement. I thought those didn't influence the package id?

Comment (SpaceIm) [reply recipes/winpcap/all/conanfile.py:80]: maybe test:

`if (hasattr(self, "settings_build") and self.settings_build.os == "Windows) or (not hasattr(self, "settings_build") and self.settings.os == "Windows")`

Comment (madebr) [reply recipes/winpcap/all/conanfile.py]: Thanks! Let's try the `settings_build` approach.

Comment (SpaceIm) [reply recipes/winpcap/all/conanfile.py]: > But it's a build requirement. I thought those didn't influence the package id?

Not related to package id. In CI of CCI, a linux machine runs some tests first, and `tools.os_info.is_windows` is always False. @jgsogo can explain better than me.

File: recipes/winpcap/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+            return self.settings_build.os
+        else:
+            return self.settings.os
+
+    def config_options(self):
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+            self.options.turbocap = True
+            self.options.capture_type = "win32"
+        elif self.settings.os == "Linux":
```
Comment (seladb) [inline recipes/winpcap/all/conanfile.py:60]: @madebr WinPcap is for Windows only. It's actually a Windows port of [libpcap](https://www.tcpdump.org/) which is not available on Windows. For Linux/macOS the [libpcap Conan package](https://conan.io/center/libpcap) should be used

Comment (madebr) [reply recipes/winpcap/all/conanfile.py:60]: Then I'll look into stripping everything non-Windows related.

File: recipes/winpcap/all/conanfile.py (hunk 4 above)
Comment (jgsogo) [reply recipes/winpcap/all/conanfile.py]: Yes. In CCI we first compute the packageIDs for all the configurations (this computation runs on a Linux machine) so we can remove duplicates and those that are invalid-config. Then, we will use win/linux/mac nodes to build the libraries. 

If (running in Linux) Conan computes an invalid-config for a given configuration (a Windows one) we won't try to build that configuration, we will just report it is invalid. For this reason, it is very important that everything used in a recipe is provided as input (via profiles) and nothing is guessed... and this is mandatory for the source lines that are executed while computed the package-id in the Linux machine (tomorrow it could be a Windows node or Macos).

Although build-requires will not affect the package-id, while computing if it is a valid configuration, we also take into account the build-requires that will be required to build. If these build-requires are not available, then the configuration is also removed.

TL;DR; `toos.os_info` is evil outside _build_ step (`build()` always takes place in the build machine).

Comment (madebr) [reply recipes/winpcap/all/conanfile.py]: Thanks for the reply.
I know you had to write this message a lot already. 😅

File: recipes/winpcap/all/patches/0001-linux-patch.cmake [hunk 6]
```diff
@@ -0,0 +1,16 @@
+--- wpcap/libpcap/pcap-linux.c
++++ wpcap/libpcap/pcap-linux.c
+@@ -129,10 +129,10 @@
+ #ifdef HAVE_REMOTE
+ #include <pcap-remote.h>
+ #endif
++#ifdef NEED_LINUX_SOCKIOS_H
```
Comment (prince-chrismc) [inline recipes/winpcap/all/patches/0001-linux-patch.cmake:7]: Bad file extension

File: recipes/winpcap/all/test_package/test_package.c (hunk 3 above)
Comment (madebr) [reply recipes/winpcap/all/test_package/test_package.c:11]: FIXME: borrow test package from https://github.com/bincrafters/community/pull/1395

File: recipes/winpcap/all/conanfile.py (hunk 4 above)
Comment (prince-chrismc) [reply recipes/winpcap/all/conanfile.py]: Should this go in the FAQ?

File: recipes/winpcap/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+            raise ConanInvalidConfiguration("missing dag cci recipe")
+        if self.options.get_safe("bluetooth"):
+            # FIXME: missing bluez recipe
+            raise ConanInvalidConfiguration("missing bluez cci recipe")
+        if self.options.turbocap and self.settings.os != "Windows":
+            # FIXME: missing turbocap recipe
+            raise ConanInvalidConfiguration("missing turbocap cci recipe")
+
+    def build_requirements(self):
+        if self._build_os == "Windows":
```
Comment (madebr) [inline recipes/winpcap/all/conanfile.py:83]: FIXME: add `_settings_build` property method

File: recipes/winpcap/all/conanfile.py (hunk 4 above)
Comment (madebr) [reply recipes/winpcap/all/conanfile.py]: See https://github.com/conan-io/hooks/pull/320 for a first attempt to implement this into a hook.

Comment (prince-chrismc) [reply recipes/winpcap/all/conanfile.py]: Amazing! this is supper helpful to have and know

**Comment by seladb** [comment]
@madebr the PR was closed :(
Any chance you'll have time to complete this work?
